prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt

```diff
@@ -25,6 +31,86 @@ import java.util.stream.Stream
+const val conditionCodeExtensionURL = "https://reportstream.cdc.gov/fhir/StructureDefinition/condition-code"
+
+/**
+ * Looks up a condition code for the passed in [code] (typically a test) using the given [metadata] object
+ * @param code test (or other type) code to look up a condition for
+ * @param metadata metadata containing an observation-mapping lookup table
+ * @return the condition code or null if no code was found
+ */
+private fun lookupCondition(code: Coding, metadata: Metadata): Coding? {
+    val mappingTable = metadata.findLookupTable("observation-mapping").also {
+        if (it == null) { // could not load the table
+            throw IllegalStateException("Unable to load lookup table 'observation-mapping' for condition stamping")
+        }
+    }!!
+    val condition = mappingTable.caseSensitiveDataRowsMap.find { // search for the code
+        it[ObservationMappingConstants.TEST_CODE_KEY] == code.code
+    }
+    return if (condition.isNullOrEmpty()) { // could not find the code
+        null
+    } else { // code found; create Coding instance to return
+        Coding(
+            condition[ObservationMappingConstants.CONDITION_CODE_SYSTEM_KEY],
+            condition[ObservationMappingConstants.CONDITION_CODE_KEY],
+            condition[ObservationMappingConstants.CONDITION_NAME_KEY]
+        )
+    }
+}
+
+/**
+ * Retrieves loinc/snomed codes from [this] observation in known locations (code.coding and valueCodeableConcept.coding)
+ * @return a map of lists of codings keyed by their origin as a printable string
+ */
+fun Observation.getCodeSourcesMap(): Map<String, List<Coding>> {
+    val toReturn = mutableMapOf<String, List<Coding>>()
+    try {
+        toReturn[ObservationMappingConstants.BUNDLE_CODE_IDENTIFIER] = this.code.coding
+    } catch (error: FHIRException) {
+        if (error.message == null ||
+            !error.message!!.startsWith("Type mismatch: the type CodeableConcept was expected")
+            ) {
+            throw error
+        }
+    }
+    try {
+        toReturn[ObservationMappingConstants.BUNDLE_CODEABLE_IDENTIFIER] = this.valueCodeableConcept.coding
+    } catch (error: FHIRException) {
+        if (error.message == null ||
+            !error.message!!.startsWith("Type mismatch: the type CodeableConcept was expected")
+            ) {
+            throw error
+        }
+    }
+    return toReturn
+}
+
+/**
+ * For every snomed/loinc code in code or valueCodeableConcept, lookup a condition code and add it as an extension
+ * @param metadata metadata containing an observation-mapping lookup table
+ * @return a list of ActionLogDetail objects with information on any mapping failures
+ */
+fun Observation.addMappedCondition(metadata: Metadata): List<ActionLogDetail> {
+    val codeSourcesMap = this.getCodeSourcesMap().filterValues { it.isNotEmpty() }
+    if (codeSourcesMap.values.flatten().isEmpty()) return listOf(UnmappableConditionMessage()) // no codes found
```

Comment:
Nit: maybe a different class here since it's a little confusing since there isn't anything to map right here?

Reply:
I'm not against it but I liked the elegance of using one class. In either case (no codes or codes without lookup values) the observation has an unmappable condition (reasoning on the name).

It also didn't make the query easier to distinguish between the two but maybe I was missing something there.

Your choice on this one either way.

Reply:
No strong feelings from me!

Reply:
After some more thought I think your suggestion is more scalable & testable so I went with that :).